commit 196bbeb0a5ba3678adc7faa2f292ecfdfc0fb061
Author: Christian Ehrhardt <christian.ehrhardt@codasip.com>
Date:   Wed Feb 12 10:54:02 2025 +0100

    cheri: Fix array printing for local arrays
    
    The print_array() family of functions is used to print both
    local and remote arrays. For local arrays the input buffer
    in strace itself must be a true pointer while for remote
    printing the buffer is just an address in the tracee. Allow
    for a pointer and pass in pointers when needed to print a
    local array and plain addresses when printing a remote array.

Upstream-Status: Inappropriate (at the moment) as Cheri specific

diff --git a/src/defs.h b/src/defs.h
index 263fb89..3f1af33 100644
--- a/src/defs.h
+++ b/src/defs.h
@@ -1092,7 +1092,7 @@ typedef void (*print_obj_by_addr_size_fn)(struct tcb *,
  */
 extern bool
 print_array_ex(struct tcb *,
-	       kernel_ulong_t start_addr,
+	       uintptr_t start_addr,
 	       size_t nmemb,
 	       void *elem_buf,
 	       size_t elem_size,
@@ -1106,7 +1106,7 @@ print_array_ex(struct tcb *,
 /** Print an array from tracee's memory without any index printing features. */
 static inline bool
 print_array(struct tcb *const tcp,
-	    const kernel_ulong_t start_addr,
+	    const uintptr_t start_addr,
 	    const size_t nmemb,
 	    void *const elem_buf,
 	    const size_t elem_size,
diff --git a/src/util.c b/src/util.c
index 92c0577..89eea29 100644
--- a/src/util.c
+++ b/src/util.c
@@ -1709,7 +1709,7 @@ print_fd_array_member(struct tcb *tcp, void *elem_buf, size_t elem_size,
  */
 bool
 print_array_ex(struct tcb *const tcp,
-	       const kernel_ulong_t start_addr,
+	       const uintptr_t start_addr,
 	       const size_t nmemb,
 	       void *elem_buf,
 	       const size_t elem_size,
@@ -1747,7 +1747,7 @@ print_array_ex(struct tcb *const tcp,
 	const kernel_ulong_t abbrev_end =
 		(abbrev(tcp) && max_strlen < nmemb) ?
 			start_addr + elem_size * max_strlen : end_addr;
-	kernel_ulong_t cur;
+	uintptr_t cur;
 	kernel_ulong_t idx = 0;
 	enum xlat_style xlat_style = flags & XLAT_STYLE_MASK;
 	bool truncated = false;
